### DISCUSSION TITLE
Pydantic as the validation layer of dataclasses

### DISCUSSION THREAD

**Discussion by gbrennon** [pydantic/pydantic#2800] [Question]
Hey folks, how are u doing?

I saw in the docs that I can generate pydantic models based on pure python dataclasses and this is amazing!

Based on this I started to think about implementing a way of generating dynamic schemas using this feature., but I'm experiencing some difficulties when trying to convert it back to my dataclass :(

How can I do this? I implemented something like this:

```python
from pydantic.dataclasses import dataclass as pydantic_dataclass
from dataclasses import dataclass

@dataclass
class Foo:
    bar: str

FooSchema = pydantic_dataclass(Foo)
validated = FooSchema(**data)
```
But now I want to convert the schema back to my dataclass! 

I tried using `dataclass.asdict` but this mess up the types... 

How can I instantiate my class with all the correct fields that pydantic did validate?
How can I do this?

PS: I need to support nested classes

**Comment by samuelcolvin** [comment]
`validated` will be a dataclass of type `FooSchema`, but it's almost identical to a dataclass of type `Foo` - the only difference is some internal dunder attribute and methods used for the validation.

You'll need to explain why you can't just use `validated` for whatever you need to do.

**Reply by gbrennon** [reply]
Hey samuel! How are u doing?
Man, I agree with u that they are almost identical! But in that case I'm afraid because my `dataclass` can have some method and 

I would lose them after validation :(

I'm trying to do this to make a, somehow, hexagonal architecture works using the `pydantic` as my `validation` between `application layer` and `api` through DTOs.

For example:
```python
# domain entity
class Product:
    def __init__(self, name: str, price: Decimal, id_: Optional[UUID] = None):
        self.__id = id_ or uuid4()
        self.__name = name
        self.__price = price

# usecase dto
@dataclass
class CreateProductInputDTO:
    name: str
    price: Decimal

    def to_entity(self) -> Product:
        return Product(self.name, self.price)

# application layer
class CreateProductUseCase:
    def __init__(self, product_repository: ProductRepository):
        self.__product_repository = product_repository

    def execute(input_dto: CreateProductInputDTO) -> CreateProductOutputDTO:
        product = input_dto.to_entity()
        self.__product_repository.save(product)
        return CreateProductOutputDTO.from_entity(product)        

# API layer
class CreateProductAPI:
    @validate_dto(CreateProductInputDTO)
    def post(self, input_dto: CreateProductInputDTO):
        self.__create_product_use_case.execute(input_dto)
```

In order to this works my `validate_dto` would need to use pydantic dataclasses to generate schemas and try to convert it back to my `original dto` as it have a methods that is needed to build my `entity`

**Reply by PrettyWood** [reply]
Hi @gbrennon
Your methods will still be available in the validated `dataclass` 🤔 I don't really understand where your issue is. Maybe you can create a working example to reproduce your issue?

Btw I did some work to refactor how the pydantic `dataclass` decorator works in https://github.com/samuelcolvin/pydantic/pull/2557 that may well be added in 1.9. The goal is to make conversion from stdlib `dataclass` to _pydantic_ `dataclass` as flowless as possible. It may help you do what you want. Feel free to try it out if you can.

**Reply by gbrennon** [reply]
sorry @PrettyWood , i dont remember exactly what i was experiencing :rofl: 
i only remember that this was related to an internal taht i was designin for a company that i used to work!

also i dont remember the decision that i did took or how was the code that i did write for this internal lib :rofl: 

ill try to re-write that and see if i experience the same issue
